src/python/pants/engine/internals/native.py

```diff
@@ -31,38 +48,74 @@
 from pants.util.memo import memoized_property
 from pants.util.meta import SingletonMetaclass
 
+RuleYield = Union[Get, Tuple[Get, ...]]
+RuleSend = TypeVar("RuleSend", Tuple[Any, ...], Exception)
+RuleReturn = Any
+RuleCoroutine = Coroutine[RuleYield, RuleSend, RuleReturn]
+
+EngineGeneratorResult = Union[
+    PyGeneratorResponseGet,
+    PyGeneratorResponseGetMulti,
+    PyGeneratorResponseBreak,
+    PyGeneratorResponseThrow,
+]
+
+
 logger = logging.getLogger(__name__)
 
 
+class IntrinsicError(Exception):
+    """Exceptions raised within intrinsic methods implemented in Rust."""
+
+
 class Externs:
     """Methods exposed from Python to Rust.
 
     TODO: These could be implemented in Rust in `externs.rs` via the cpython API.
     """
 
-    def __init__(self, lib):
+    def __init__(self, lib: ModuleType):
         self.lib = lib
 
     _do_raise_keyboardinterrupt = bool(os.environ.get("_RAISE_KEYBOARDINTERRUPT_IN_EXTERNS", False))
 
-    def is_union(self, input_type):
+    def is_union(self, input_type: Type) -> bool:
         """Return whether or not a type is a member of a union."""
         # NB: This check is exposed for testing error handling in CFFI methods. This code path should
         # never be active in normal pants usage.
         return union.is_instance(input_type)
 
-    def create_exception(self, msg):
+    def create_exception(self, msg: str) -> IntrinsicError:
         """Given a utf8 message string, create an Exception object."""
-        return Exception(msg)
+        return IntrinsicError(msg)
+
+    @staticmethod
+    def _do_send(func: RuleCoroutine, arg: Tuple[Any, ...]) -> RuleYield:
+        return func.send(arg)
+
+    def generator_send(self, func: RuleCoroutine, arg: Tuple[Any, ...]) -> EngineGeneratorResult:
+        return self._generator_send_or_throw_impl(func, arg, self._do_send)
 
-    def generator_send(
-        self, func, arg
-    ) -> Union[PyGeneratorResponseGet, PyGeneratorResponseGetMulti, PyGeneratorResponseBreak]:
+    @staticmethod
+    def _do_throw(func: RuleCoroutine, arg: Exception) -> RuleYield:
```

Comment:
Same with `_do_throw`, seems like we could inline this and save the trouble of defining a new function